File: src-tauri/src/commands/credentials/broker.rs
//! Tauri commands for the Zero-Plaintext Credential Broker.
//!
//! The Broker lets external processes (fleet sessions, scripts, MCP clients)
//! *use* vault credentials without ever seeing them: they hold short-lived,
//! narrowly-scoped `external_api_keys` handles and route calls through the
//! audited `/api/proxy/{credential_id}` route. These commands power the vault
//! Broker surface: mint handles, list consumers + their observed activity,
//! and the per-consumer kill-switch.

use std::sync::Arc;
use tauri::State;

use crate::db::models::{ApiKeyAuditEntry, BrokerConsumerView, CreateApiKeyResponse};
use crate::db::repos::resources::api_key_audit;
use crate::db::repos::resources::broker_edges;
use crate::db::repos::resources::external_api_keys as api_key_repo;
use crate::db::repos::resources::settings_audit_log;
use crate::engine::credential_broker;
use crate::error::AppError;
use crate::AppState;
use personas_macros::requires;

/// Mint a short-lived derived handle for one credential. Returns the handle
/// plaintext exactly once — the credential's secret is never part of the
/// response. TTL is clamped server-side (5 min .. 24 h, default 60 min).
#[tauri::command]
#[requires(privileged)]
pub fn mint_credential_handle(
    state: State<'_, Arc<AppState>>,
    credential_id: String,
    consumer_name: String,
    ttl_minutes: Option<u32>,
) -> Result<CreateApiKeyResponse, AppError> {
    credential_broker::mint_derived_handle(&state.db, &credential_id, &consumer_name, ttl_minutes)
}

/// List all observed broker consumers (one row per consumer key), aggregated
/// across credentials and joined with live key state for the kill-switch.
#[tauri::command]
#[requires(privileged)]
pub fn list_broker_consumers(
    state: State<'_, Arc<AppState>>,
) -> Result<Vec<BrokerConsumerView>, AppError> {
    broker_edges::list_consumers(&state.db)
}

/// Recent management-API request trail for one consumer key (newest first).
/// Reuses the per-key audit table the middleware writes on every request.
#[tauri::command]
#[requires(privileged)]
pub fn list_broker_consumer_activity(
    state: State<'_, Arc<AppState>>,
    consumer_key_id: String,
    limit: Option<u32>,
) -> Result<Vec<ApiKeyAuditEntry>, AppError> {
    api_key_audit::list_for_key(&state.db, &consumer_key_id, limit.unwrap_or(50))
}

/// Kill-switch: revoke a consumer key. Takes effect on the consumer's very
/// next request (`find_by_token` filters revoked keys), and the consumer's
/// blast-radius edges drop out of the dependents graph immediately.
#[tauri::command]
#[requires(privileged)]
pub fn revoke_broker_consumer(
    state: State<'_, Arc<AppState>>,
    consumer_key_id: String,
) -> Result<(), AppError> {
    api_key_repo::revoke(&state.db, &consumer_key_id)?;
    tracing::info!(consumer_key_id = %consumer_key_id, "broker consumer revoked (kill-switch)");
    // Settings → History feed; best-effort.
    if let Err(e) = settings_audit_log::insert(
        &state.db,
        "api_keys",
        &consumer_key_id,
        "broker_kill_switch",
        None,
        None,
        Some("ui"),
    ) {
        tracing::warn!(error = %e, "settings_audit_log insert failed for broker kill-switch");
    }
    Ok(())
}

File: src/api/credentials/broker.ts
/**
 * Zero-Plaintext Credential Broker — frontend API wrappers.
 *
 * External consumers (fleet sessions, scripts, MCP clients) never hold vault
 * secrets: they hold short-lived, narrowly-scoped handles and route real API
 * calls through the audited `/api/proxy/{credential_id}` route. This module
 * powers the vault Broker surface: consumer list, per-consumer activity,
 * handle minting, and the kill-switch.
 *
 * NOTE on types: the Rust structs (`BrokerConsumerView`, `ApiKeyAuditEntry`,
 * `CreateApiKeyResponse`) are ts-rs exported; the interfaces below mirror
 * their camelCase serialization so this module does not depend on a bindings
 * regeneration step.
 */
import { invokeWithTimeout as invoke } from '@/lib/tauriInvoke';

/** One observed broker consumer (an external API key that used credentials). */
export interface BrokerConsumer {
  consumerKeyId: string;
  consumerName: string;
  /** Safe display prefix (e.g. "pk_a1b2c3"); null when the key row is gone. */
  keyPrefix: string | null;
  /** True when the key still authenticates (enabled, not revoked). */
  active: boolean;
  revokedAt: string | null;
  expiresAt: string | null;
  credentialIds: string[];
  credentialNames: string[];
  totalCalls: number;
  lastStatus: number | null;
  lastUsedAt: string | null;
}

/** One management-API request a consumer key made (per-key audit trail). */
export interface BrokerConsumerActivityEntry {
  id: string;
  key_id: string;
  at: string;
  method: string;
  path: string;
  status: number;
  persona_id: string | null;
  origin: string | null;
}

/** Response from minting a derived handle — plaintext returned exactly once. */
export interface MintedHandle {
  record: {
    id: string;
    name: string;
    key_prefix: string;
    scopes: string;
    enabled: boolean;
    created_at: string;
    expires_at: string | null;
    label: string | null;
  };
  plaintext_token: string;
}

/** All observed broker consumers, most recently active first. */
export const listBrokerConsumers = () =>
  invoke<BrokerConsumer[]>('list_broker_consumers');

/** Recent request trail for one consumer key (newest first). */
export const listBrokerConsumerActivity = (consumerKeyId: string, limit?: number) =>
  invoke<BrokerConsumerActivityEntry[]>('list_broker_consumer_activity', {
    consumerKeyId,
    limit,
  });

/**
 * Mint a short-lived derived handle for one credential. The returned plaintext
 * is shown once and never retrievable again; the credential's secret is never
 * part of the response. TTL clamps server-side to 5 min .. 24 h (default 60).
 */
export const mintCredentialHandle = (
  credentialId: string,
  consumerName: string,
  ttlMinutes?: number,
) =>
  invoke<MintedHandle>('mint_credential_handle', {
    credentialId,
    consumerName,
    ttlMinutes,
  });

/**
 * Kill-switch: revoke a consumer key. Effective on the consumer's next
 * request; its blast-radius edges drop from the dependents graph immediately.
 */
export const revokeBrokerConsumer = (consumerKeyId: string) =>
  invoke<void>('revoke_broker_consumer', { consumerKeyId });
